src/schema/me/__tests__/followed_shows.test.ts

```diff
@@ -0,0 +1,87 @@
+/* eslint-disable promise/always-return */
+import { runAuthenticatedQuery } from "test/utils"
+import gql from "lib/gql"
+import cityData from "../../../schema/city/cityDataSortedByDisplayPreference.json"
+import { LOCAL_DISCOVERY_RADIUS_KM } from "../../../schema/city/constants"
+
+const BASE_GRAVITY_ARGS = { size: 10, offset: 0, total_count: true }
+const stubResolver = () => Promise.resolve({ body: [], headers: {} })
+
+const generate_query = (params = `(first: 10)`) =>
+  gql`
+    {
+      me {
+        followsAndSaves {
+          shows${params} {
+            edges {
+              node {
+                id
+              }
+            }
+          }
+        }
+      }
+    }
+  `
+
+describe("returns followed shows for a user", () => {
+  let followedShowsLoader
+
+  beforeEach(() => {
+    followedShowsLoader = jest.fn(stubResolver)
+  })
+
+  it("generates a predictable URL with no parameters", async () => {
+    const query = generate_query()
+    await runAuthenticatedQuery(query, { followedShowsLoader })
+    expect(followedShowsLoader).toHaveBeenCalledWith(BASE_GRAVITY_ARGS)
+  })
+
+  describe("respects optional city parameter", () => {
+    const generate_city_args = city =>
+      !city
+        ? BASE_GRAVITY_ARGS
+        : {
+            ...BASE_GRAVITY_ARGS,
+            near: `${city.coordinates.lat},${city.coordinates.lng}`,
+            max_distance: LOCAL_DISCOVERY_RADIUS_KM,
+          }
+
+    const buildTestData = city => ({
+      name: city.name,
+      query: generate_query(`(first: 10, city: "${city.slug}")`),
+      expectedArgs: generate_city_args(city),
+    })
+
+    cityData.map(buildTestData).forEach(({ name, query, expectedArgs }) => {
+      // I totally grant that this is a bit weird
+      // but the behavior under test is tightly coupled
+      // to the contents of the cityData file. If a change
+      // to that file breaks this test I want to know about it,
+      // and this will tell us.
+
+      // When Jest hits this point in the code it will treat this almost
+      // like a macro - we'll have a new test generated for each city,
+      // you'll see the test results enumerated as if we'd done this manually.
+      it.call(null, `Handles City: ${name}`, async () => {
```

Comment:
Yeah, I'm torn here.

An existing and probably simpler to grok pattern is to simply [mock the contents of the city data file](https://github.com/anandaroop/metaphysics/blob/ec1f87ef91feb8924082c9ad7fdd51a09f778417/src/schema/__tests__/city.test.ts#L1-L25) for the purposes of this test. I think that makes sense because we don't care so much about the literal contents of this file, just the logic that makes use of those contents. (Having six specs as opposed to one here, doesn't actually _increase_ our confidence I don't think.)

The `it.call` totally threw me  — I went looking for this in the Jest api before I realized you're just calling `call` on a function, which is what `it` is.

I think I'm coming down slightly on the "too-clever" side (though now that I know how it works, the `it.call` thing is neat trick to know about).

Reply:
So, the mock approach exactly doesn't solve the thing I'm caring about. If we mock the city data for the purposes of this test then future edits to the shape and structure of that won't be reflected in the actual test here, right? That's why this is confusing me so much, because literally the behavior of the function depends on the specific data that currently lives inside of that city data file, you know? If we test a mock, and the mock diverges, then our tests don't give us any confidence as to whether or not we're actually accurately representing the city's parameters in the query.

Am I just overthinking this?

Reply:
It's true, the shape of the data file could change under your feet and then these tests that rely on mocked data wouldn't catch that. In that sense this test is as brittle as one that mocked out data from a remote service — basically every other test in this repo 😅.

I guess I'm counting on anyone who redesigns the shape of that file to simply check where else the file is used and make the concomitant changes? Exactly that has [already happened once](https://github.com/anandaroop/metaphysics/commit/0ac1bf8f05ee32a305a8b619263780ce1af6aab7).

But your version is admittedly more paranoid, and usually _I'm_ the paranoid one, so I don't know what to believe anymore 😛.

Reply:
> If we mock the city data for the purposes of this test then future edits to the shape and structure of that won't be reflected in the actual test here, right?

I definitely hear you here. To address this concern, we could also have a dedicated test file for `cityList.ts`. That way, the complexity of `it.call` can be isolated while still making sure that the data adheres to our expectations. The unit tests remain single-purpose, and we can implement Roop's suggestion of mocking out the file for this test. Thoughts?

Reply:
Just pushed an update which:

- makes an assert against the first city in cityData
- asserts that cityData matches a snapshot
- includes a comment on that snapshot test encouraging reflection on cross-dependencies between these files.

Does this work?